### PR TITLE
Fix memory leaks in netebpfext

### PR DESCRIPTION
**Pull request by saxena-anurag**
## Description

This PR does the following:
1. Enable memory leak detection for `fault_injection_netebpfext` tests.
2. Fix memory leaks discovered by the test.
3. Remove `unit_tests_leak_detection`, `netebpfext_unit_leak_detection` and enable `leak_detection` in `unit_tests`, `netebpf_ext_unit_tests`.

## Testing

This PR adds the memory leak detection test.

## Documentation

NA

### CODE REVIEW

File: netebpfext/net_ebpf_ext_hook_provider.c [hunk 1]
```diff
@@ -368,8 +368,10 @@ _net_ebpf_extension_hook_provider_attach_client(
         *provider_binding_context = hook_client;
         hook_client = NULL;
     } else {
-        if (hook_client)
```
Comment (shankarseal) [inline netebpfext/net_ebpf_ext_hook_provider.c:371]: I would rather have a function to clean up hook_client including fields such as work items.

File: netebpfext/net_ebpf_ext_sock_ops.c [hunk 2]
```diff
@@ -203,42 +203,35 @@ _net_ebpf_extension_sock_ops_on_client_detach(_In_ const net_ebpf_extension_hook
...
-        while (!IsListEmpty(&filter_context->flow_context_list.list_head)) {
-            LIST_ENTRY* entry = RemoveHeadList(&filter_context->flow_context_list.list_head);
-            InitializeListHead(entry);
-            net_ebpf_extension_sock_ops_wfp_flow_context_t* flow_context =
-                CONTAINING_RECORD(entry, net_ebpf_extension_sock_ops_wfp_flow_context_t, link);
-            flow_context->client_detached = TRUE;
-            _Analysis_assume_(flow_index < flow_count);
-            flow_parameters_array[flow_index++] = flow_context->parameters;
-        }
+    if (filter_context->flow_context_list.count > 0) {
```
Comment (shankarseal) [inline netebpfext/net_ebpf_ext_sock_ops.c:214]: was the intent to not have allocation within the lock? Where was the leak here?

File: netebpfext/net_ebpf_ext_bind.c [hunk 3]
```diff
@@ -126,6 +126,11 @@ _net_ebpf_extension_bind_on_client_attach(
         (net_ebpf_extension_hook_client_t*)attaching_client, filter_context);
 
 Exit:
+    if (result != EBPF_SUCCESS) {
```
Comment (shankarseal) [inline netebpfext/net_ebpf_ext_bind.c:129]: Maybe call `net_ebpf_extension_wfp_filter_context_cleanup` instead?

File: netebpfext/net_ebpf_ext_sock_ops.c (hunk 2 above)
Comment (saxena-anurag) [reply netebpfext/net_ebpf_ext_sock_ops.c:214]: This function was allocating an array in the client detach path. If the allocation failed, this function failed to call `FwpsFlowRemoveContext` for all the flows and returned an error. That caused memory leak.

While fixing, I realized we dont need to actually allocate memory and can also snap the list under lock and later iterate over the list.

**Review by dthaler — APPROVED** [review]

File: netebpfext/net_ebpf_ext_bind.c (hunk 3 above)
Comment (saxena-anurag) [reply netebpfext/net_ebpf_ext_bind.c:129]: Yes we can call `net_ebpf_extension_wfp_filter_context_cleanup`, but all other instances (in files sock_addr.c, sock_ops.c, etc.) are directly calling `ExFreePool`, so I used the same here for parity.

File: netebpfext/net_ebpf_ext_hook_provider.c (hunk 1 above)
Comment (saxena-anurag) [reply netebpfext/net_ebpf_ext_hook_provider.c:371]: fixed.

File: netebpfext/net_ebpf_ext_hook_provider.c [hunk 4]
```diff
@@ -284,6 +284,20 @@ net_ebpf_extension_hook_check_attach_parameter(
     NET_EBPF_EXT_RETURN_RESULT(result);
 }
 
+void
+_net_ebpf_extension_hook_client_cleanup(_Frees_ptr_opt_ net_ebpf_extension_hook_client_t* hook_client)
+{
+#pragma warning(push)
+#pragma warning(disable : 6001) // Using uninitialized memory '*hook_client'.
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext_hook_provider.c:291]: Does this warning go way if you add the `_In_opt_` annotation to line 288?  I suspect that's the right fix.

**Review by dthaler — APPROVED** [review]

File: netebpfext/net_ebpf_ext_hook_provider.c (hunk 4 above)
Comment (saxena-anurag) [reply netebpfext/net_ebpf_ext_hook_provider.c:291]: Yes that worked! Have pushed another commit.

But I am not sure why is it not giving error for (many) other similar instances. Example:
https://github.com/microsoft/ebpf-for-windows/blob/bebb9a14cd406d15bc312f12f5b1e628602d5647/netebpfext/net_ebpf_ext_prog_info_provider.c#L116

**Review by dthaler — APPROVED** [review]

**Review by Alan-Jowett — APPROVED** [review]

**Pull request merged**